src/snowflake/connector/storage_client.py

```diff
@@ -445,9 +445,10 @@ def _abort_multipart_upload(self) -> None:
     def delete_client_data(self) -> None:
         """Deletes the tmp_dir and closes the source stream belonging to this client.
         This function is idempotent."""
-        if os.path.exists(self.tmp_dir):
-            logger.debug(f"cleaning up tmp dir: {self.tmp_dir}")
-            shutil.rmtree(self.tmp_dir)
+        with self.lock:
+            if os.path.exists(self.tmp_dir):
+                logger.debug(f"cleaning up tmp dir: {self.tmp_dir}")
+                shutil.rmtree(self.tmp_dir)
```

Comment:
I think it'd be acceptable here to just catch all exceptions coming out of `shutil.rmtree` and ignore them. And so if 2 threads are cleaning up simultaneously then if 1 of them is already cleaning up the other one could just give up on doing the same thing.